src/sentry/outcome_consumer.py

```diff
@@ -0,0 +1,108 @@
+"""
+The OutcomeConsumer is a task that runs a loop in which it reads outcome messages coming on a kafka queue and
+processes them.
+
+Long Story: Event outcomes are placed on the same Kafka event queue by both Sentry and Relay.
+When relay generates an outcome for a message it also sends a signal ( a Django signal) that
+is used by getSentry for internal accounting.
+
+Relay (running as a Rust process) cannot send django signals so in order to get outcome signals sent from
+Relay into getSentry we have this outcome consumers which listens to all outcomes in the kafka queue and
+for outcomes that were sent from Relay sends the signals to getSentry.
+
+In conclusion the OutcomeConsumer listens on the the outcomes kafka topic, filters the outcomes by dropping
+the outcomes that originate from sentry and keeping the outcomes originating in relay and sends
+signals to getSentry for these outcomes.
+
+"""
+from __future__ import absolute_import
+
+import logging
+
+from django.conf import settings
+
+from sentry.models.project import Project
+from sentry.signals import event_filtered, event_dropped
+from sentry.utils.kafka import SimpleKafkaConsumer
+from sentry.utils import json
+from sentry.utils.outcomes import is_outcome_signal_sent, mark_outcome_signal_sent, Outcome
+
+logger = logging.getLogger(__name__)
+
+
+class OutcomesConsumer(SimpleKafkaConsumer):
+    def process_message(self, message):
+        msg = json.loads(message.value())
+
+        project_id = msg.get("project_id")
+
+        if not project_id:
+            return  # no project
+        project_id = int(project_id)
+
+        event_id = msg.get("event_id")
+
+        if is_outcome_signal_sent(project_id=project_id, event_id=event_id):
+            return  # message already processed nothing left to do
+
+        outcome = int(msg.get("outcome", -1))
+        reason = msg.get("reason")
+        remote_addr = msg.get("remote_addr")
+
+        if outcome == Outcome.FILTERED or outcome == Outcome.RATE_LIMITED:
+            # try to get the project
+            try:
+                project = Project.objects.get_from_cache(id=project_id)
+            except Project.DoesNotExist:
+                return
```

Comment:
Please log an error here. This is a condition that should never happen and might indicate misconfiguration or a serious bug.

Reply:
good point, done